src/dbnode/encoding/proto/README.md

```diff
@@ -0,0 +1,269 @@
+
+Every write is prefixed with a control bit.
+
+If the control bit is set to `1` then we know that the stream contains another write that we need to decode and we can begin decoding the timestamp.
+
+If the control bit is set to `0` then we have either: **a)** reached the end of the stream *or* **b)** encountered a time unit change.
+
+We resolve the ambiguity by reading the next control bit which will be `0` if we've reached the end of the stream or `1` if we should interpret the next 8 bits as the new time unit that we should use for delta of delta timestamp encoding.
+
+We have to keep track of time unit changes on our own instead of deferring to the M3TSZ delta-of-delta encoder which can handle this independently because the M3TSZ encoder relies on a custom marker scheme to indicate time unit changes.
+It avoids using a control bit for each write by using a marker which contains a prefix that could not possibly be generated into the stream by any possible input and then the decoder frequently "looks ahead" for this marker to see if it needs to decode a time unit change.
+The Protobuf encoding scheme has no equivalent "impossible bit combination" so it uses explicit control bits to indicate a time unit change instead.
+
+### Compressed Timestamp
+
+The Protobuf compression scheme reuses the delta of delta timestamp encoding logic that is implemented in the M3TSZ package and decribed in the [Facebook Gorilla paper](https://www.vldb.org/pvldb/vol8/p1816-teller.pdf).
+
+
+After encoding a control bit with a value of `1` indicating that there is another write in the stream, we encode the delta of delta of the current timestamp and the previous one as described in the Gorilla paper.
+
+Similarly, when we're decoding the stream, we perform the inverse operation and reconstruct the current timestamp based on the previous one and the delta of delta encoded into the stream.
+
+### Compressed Protobuf Fields
+
+Compressing the protobuf fields is broken into two stages:
+
+1. Custom compressed fields
+2. Protobuf marshaled fields
+
+In the first phase we compress any fields for which we're able to apply custom compression as described in the `Compression Techniques` section.
+
+In the second phase we lean on the Protobuf marshaling format to encode the data for us with the caveat that we compare fields at the top-level and avoid re-encoding them if they haven't changed.
+
+#### Custom Compressed Protobuf Fields
+
+We encode the custom compressed field values similarly how to we encode their types as described in the `Header` section.
+In fact, they're even encoded in the same order with the caveat that unlike when we're encoding the types, we don't need to encode a null value for non-contiguous field numbers for which we're not performing any compression.
```

Comment:
> In fact, they're encoded in the same order, with the caveat that unlike when types are being encoded, a null value does not need to be encoded for non-contiguous field numbers that are not being compressed.

also, jfmi, does this mean:

- non-contiguous (has gaps);
- unordered; or
- non-sequential

?

Reply:
has gaps if I understand your question correctly